### PR TITLE
Quantum Kitchen Sinks

### PR DESCRIPTION
**Pull request by zackgow**
This is initial code to reproduce [Quantum Kitchen Sinks](https://arxiv.org/pdf/1806.08321.pdf). This shows that we can successfully apply the QKS transformation to the picture frames dataset. The classification accuracy is over 99% with logistic regression once the transformation is applied.

### CODE REVIEW

File: Experiments/forest_QKS/report.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+## Data
+The synthetic "picture frames" dataset proposed in the paper was generated via the `make_frames` function. This creates a smaller inner square with sides of length 2 and a larger outer square with sides of length 4. Both have points uniformly distributed around the average. The training set is composed of 1600 two-dimensional points, 800 for each class. While the test set is 400 two-dimensional points, 200 for each class. 
+
+
+picture frames             |  picture frames
+:-------------------------:|:-------------------------:
+![training set](https://github.com/zackgow/public_research/raw/quantum_kitchen_sinks/Experiments/forest_QKS/figs/training_dataset.png) |  ![test set](https://github.com/zackgow/public_research/raw/quantum_kitchen_sinks/Experiments/forest_QKS/figs/test_dataset.png)
+
+## Results
+Using the parameters below, we were able to achieve an accuracy of **0.995%** on the training set and **0.9975%** on the test set using logistic regression after obtaining the feature vector from the Quantum Kitchen Sinks method.
```
Comment (mstechly) [inline Experiments/forest_QKS/report.md]: This accuracy is extremely low! Under 1%?
Please change it to 99.5%  ;)

File: Experiments/forest_QKS/qks.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+    of length 4. Both have points uniformally distributed around
+    the average.
+    """
+    def zip_arrays(d1,d2):
+        return np.array(list(zip(d1,d2)))
+
+    def make_squares(num, l):
+        left_right, top_bottom = make_barriers(num,l), make_barriers(num,l)
+        top_bottom[:,[0,1]] = top_bottom[:,[1,0]] 
+        square = np.concatenate((left_right, top_bottom))
```
Comment (mstechly) [inline Experiments/forest_QKS/qks.py]: Your placement of spaces after commas is inconsistent. I recommend that you always put a space after a comma or arithmetic operators.
See: https://stackoverflow.com/questions/18171560/does-pep-8-require-whitespace-around-operators-in-function-arguments

File: Experiments/forest_QKS/qks.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+        return barriers
+    
+    outer = make_squares(num, l=2)
+    inner = make_squares(num, l=1)
+    frames = np.concatenate((outer,inner))
+    
+    half = int((len(frames)/2))
+    target = np.concatenate((np.zeros(half),np.ones(half)))
+    
+    X,y = shuffle(frames, target, random_state=1337)
```
Comment (mstechly) [inline Experiments/forest_QKS/qks.py]: I don't see why X is capital and y is not :)

File: Experiments/forest_QKS/qks.py [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+    return X,y
+
+
+def make_omega(n_episodes,p,q,r,scale,loc=0):
```
Comment (mstechly) [inline Experiments/forest_QKS/qks.py]: I have no idea what omega, p, q and r mean.
It would be great if you could find a more descriptive name. 
If not (e.g. that's what they used in the paper and there is no easy interpretation) please at least add some explanation in the function description.

Comment (mstechly) [reply Experiments/forest_QKS/qks.py]: Actually I also don't know what "loc" is ;)

File: Experiments/forest_QKS/qks.py [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+        Generates a matrix of 0s and 1s to zero out `r` values per matrix
+        """
+        matrix_size = p*q
+        m = np.zeros(matrix_size)
+        for i in range(r): m[i] = 1
+        np.random.shuffle(m)
+        selection_matrix = m.reshape(q,p)
+        return selection_matrix
+
+    norm_dist = np.random.normal(loc, scale, size=(n_episodes,q,p))
```
Comment (mstechly) [inline Experiments/forest_QKS/qks.py]: norm means "normal" here, and not "normalized" right?

File: Experiments/forest_QKS/qks.py [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+    This is probably best just left up to the compiler.
+
+    For ease of reading the printed operations, the qubits are looped over several times.
+    """    
+    program = Program()
+    qubits = qc.qubits()
+    n_qubits = len(qubits)
+    var = 'theta'
+
+    ro = program.declare('ro', memory_type='BIT', memory_size=n_qubits)
```
Comment (mstechly) [inline Experiments/forest_QKS/qks.py]: If ro is for greek letter rho, then please change it to rho.
Otherwise, I have no idea what it means.

File: Experiments/forest_QKS/qks.py [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+    thetas = {var+str(qubit):program.declare(var+str(qubit), memory_type = 'REAL') for qubit in qubits}
+
+    sq = int(np.sqrt(n_qubits))
+    lim = n_qubits-sq-1 
+    
+    for m in qubits: program += (RX(thetas[var+str(m)],m))
```
Comment (mstechly) [inline Experiments/forest_QKS/qks.py]: I understand it's a one-line for loop syntax.
Please change it to two lines - it's much easier to read then.

File: Experiments/forest_QKS/qks.py [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+    for m in qubits:
+        m_1 = m+1
+        m_sq = m+sq
+        skip = (m_1)%sq
+        
+        if m_1 < n_qubits:
+            if (m_sq >= n_qubits): program += CNOT(m,m_1)
+            else: program += CNOT(m,m_sq)
+    
+        if (m < lim) and (skip != 0): program += CNOT(m,m_1)
```
Comment (mstechly) [inline Experiments/forest_QKS/qks.py]: I see you use this "one-line" syntax a lot.
Personally, I think it's a little bit confusing and doesn't add anything to the code. 
So if you don't mind, please change all such places to regular syntax.

File: Experiments/forest_QKS/qks.py [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+    
+    img_path = 'figs/'
+    num_cols = n_episodes*q # number of columns after QKS transformation
+
+    # make picture frames dataset
+    train, y_train = make_frames(train_size)
+    test, y_test = make_frames(test_size)
+
+    p = train.shape[-1] # dimension of input vector
+    r = 1 if p/q < 1 else int(p/q)  # number of elements that are non-zero s.t. r<=p
```
Comment (mstechly) [inline Experiments/forest_QKS/qks.py]: This (ternary operator) is actually one of the few places where I think "one-lining" makes sense. 
1. Since the logic is super easy
2. Since ternary operators are common in different languages, so people are used to it.

File: Experiments/forest_QKS/qks.py (hunk 3 above)
Comment (zackgow) [reply Experiments/forest_QKS/qks.py]: Naming (data, target) as (X, y) is common and is convention in scikit-learn. I can rename to (data, target) if you'd prefer.

File: Experiments/forest_QKS/qks.py (hunk 4 above)
Comment (zackgow) [reply Experiments/forest_QKS/qks.py]: The name is true to what's in the paper, so I think `omega` should still be used. I'll leave a more informative docstring on how to interpret this. `p`, `q`, and `r` I define with comments in `main()`, but I can include descriptions in the functions as well.

File: Experiments/forest_QKS/qks.py (hunk 5 above)
Comment (zackgow) [reply Experiments/forest_QKS/qks.py]: Yes, I'll rename for clarification.

File: Experiments/forest_QKS/qks.py (hunk 3 above)
Comment (mstechly) [reply Experiments/forest_QKS/qks.py]: Since it's sci-kit convention let it be. I just don't understand why it's not (X, Y)

File: Experiments/forest_QKS/qks.py (hunk 6 above)
Comment (mstechly) [reply Experiments/forest_QKS/qks.py]: Then I think "readout" will be a much better name. WDYT?

Comment (zackgow) [reply Experiments/forest_QKS/qks.py]: I agree. However, every example from the documentation uses `ro`. So a case can be made either way.
